src/osd/PG.cc

```diff
@@ -4011,6 +4027,16 @@ void PG::chunky_scrub(ThreadPool::TPHandle &handle)
 	  scrubber.reserved_peers.clear();
 	}
 
+	{
+	  ObjectStore::Transaction t;
+	  if (scrubber.store) {
+	    scrubber.store->cleanup(&t);
+	  }
```

Comment:
This is actually a problem.  If there are still any in flight operations on the scrubber.store from a previous scrub, resetting it here will cause those transactions to access free'd memory.  I think that the solution is that whenever you need to throw away the scrubber.store, you need to:
1) Pass a transaction to a cleanup operation.
2) The cleanup operation should only delete the object, it should not perform any operations using the MapCacher (we don't want any completions on this transaction to touch any scrubber.store state).
3) Outside of cleanup, you should register_on_complete a Context which owns the scrubber.store unique_ptr (you can std::move it).
4) That leaves scrubber.store NULL, and ready to be replaced with a newly constructed cache and backing temp object while ensuring that the old one survives until all callbacks which reference it have completed.